### PR TITLE
[#174860845] Add bonus-redeem

### PR DESCRIPTION
**Pull request by pasqualedevita**
Script to reprocess redeemed requests.

How to use (config **prod_bonus**, select **option 4**)
```./bin/run bonuses-bv:redeem-bonuses --day=DAY --fromTime=FROM_TIME --toTime=TO_TIME --apiKey=API_KEY```

Test command (config **dev**, select **option 2**)
```./bin/run bonuses-bv:redeem-bonuses --day=2020-09-23 --apiKey=API_KEY --apiUrl=https://httpbin.org/anything```

### CODE REVIEW

**Comment by gunzip** [comment]
it's not clear if this command does both thing (download + reprocess) anyway it's called (the description inside is "Download redeemed bonuses"). can we 1) put a better description 2) eventually split this in two different commands if download-only is really needed?

File: src/commands/bonuses-bv/redeem-bonuses.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            "Content-Type": "application/json"
+          },
+          body: JSON.stringify(rawDataRequest)
+        });
+        const content = await rawDataResponse.json();
+        outputResults.push({
+          file: redeemedRequest,
+          statusCode: rawDataResponse.status.toString()
+        });
+        // pause 5 seconds
```
Comment (gunzip) [inline src/commands/bonuses-bv/redeem-bonuses.ts:105]: why? (can you add a comment?)

File: src/commands/bonuses-bv/redeem-bonuses.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+          file: {
+            header: "file"
+          },
+          statusCode: {
+            header: "statusCode"
+          }
+        },
+        {
+          printLine: this.log,
+          ...parsedFlags // parsed flags
```
Comment (gunzip) [inline src/commands/bonuses-bv/redeem-bonuses.ts:122]:
```suggestion
          ...parsedFlags
```

File: src/utils/azure.ts [hunk 3]
```diff
@@ -26,35 +30,77 @@ export const agid: IAzureConfig = {
...
   storageMessagesContainer: "message-content",
+  storageBonusRedeemedContainer: "",
   storageName: "agidstoragetest"
 };
 
 export const dev: IAzureConfig = {
-  configName: "io-dev-aks-k8s-01",
-  cosmosDatabaseName: "io-dev-sqldb-db-01",
+  configName: "io-d",
+  cosmosDatabaseName: "raimondo",
```
Comment (gunzip) [inline src/utils/azure.ts:45]: what? :)

File: src/utils/azure.ts [hunk 4]
```diff
@@ -26,35 +30,77 @@ export const agid: IAzureConfig = {
...
-  resourceGroup: "io-dev-rg",
+  cosmosSenderServicesContainer: "sender-services",
+  cosmosUserDataProcessingContainer: "user-data-processing",
+  cosmosUserBonusesContainer: "user-bonuses",
+  cosmosName: "io-d-cosmos-free",
+  resourceGroup: "io-d-rg-common",
   storageMessagesContainer: "message-content",
-  storageName: "iodevsaappdata"
+  storageBonusRedeemedContainer: "redeemed-test",
+  storageName: "devpasqualesa"
```
Comment (gunzip) [inline src/utils/azure.ts:59]: ?

**Comment by pasqualedevita** [comment]
> it's not clear if this command does both things (download + reprocess) anyway it's called (the description inside is "Download redeemed bonuses"). can we 1) put a better description 2) eventually split this in two different commands if download-only is really needed?

Thank you, I've fixed the description.
The first implementation is designed with two commands (download and reprocess) but I also think is better to have one command.

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 1 above)
Comment (pasqualedevita) [reply src/commands/bonuses-bv/redeem-bonuses.ts]: changed to 1 second
added comment: pause 1 second to avoid 429 statusCode response

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 2 above)
Comment (pasqualedevita) [reply src/commands/bonuses-bv/redeem-bonuses.ts]: fixed, removed comment

File: src/utils/azure.ts (hunk 3 above)
Comment (pasqualedevita) [reply src/utils/azure.ts]: We finished free cosmos containers :) Anyway I fixed it with standard `db` container.
I think that `configName` is never used.

File: src/utils/azure.ts (hunk 4 above)
Comment (pasqualedevita) [reply src/utils/azure.ts]: fixed and created `iodsttest` in dev environment

File: src/commands/bonuses-bv/redeem-bonuses.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+      cli.action.start("Sending redeemed requests started...");
+      // Send redeemed requests
+      for (const redeemedRequest of redeemedRequests) {
+        this.log(redeemedRequest);
+        const rawDataRequest = fs.readFileSync(
+          redeemedRequest,
+          "utf8"
+        ) as string;
+        const rawDataResponse = await fetch(parsedFlags.apiUrl, {
```
Comment (francescopersico) [inline src/commands/bonuses-bv/redeem-bonuses.ts:95]: You are going to flood the server with a large amount of requests that contains a lot of redeemed bonus.
I think it can be useful to have a little delay between each request.

File: src/utils/azure.ts [hunk 6]
```diff
@@ -11,9 +11,13 @@ export interface IAzureConfig {
   cosmosNotificationContainer: string;
   cosmosNotificationStatusContainer: string;
   cosmosServicesContainer: string;
+  cosmosSenderServicesContainer: string;
+  cosmosUserDataProcessingContainer: string;
```
Comment (francescopersico) [inline src/utils/azure.ts:15]: Are these related to this PR?

**Review by francescopersico — CHANGES_REQUESTED** [review]

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 5 above)
Comment (francescopersico) [reply src/commands/bonuses-bv/redeem-bonuses.ts:95]: Just noticed there is already a 1sec delay.

Comment (francescopersico) [reply src/commands/bonuses-bv/redeem-bonuses.ts:95]: Just noticed that there is already a 1sec delay.
I think it is safer to set a 5/10sec delay.

**Comment by pasqualedevita** [comment]
1. reprocess redeemed requests without save requests as local file
1. setted pause to 10s
1. removed required attribute to flags with default value
1. renamed flag containerFolder in day
1. fixed descriptions and comments

File: README.md [hunk 7]
```diff
@@ -199,6 +200,25 @@ EXAMPLE
...
+## `io-ops bonuses-bv:redeem-bonuses`
+
+Download and reprocess redeemed bonuses
+
+```
+USAGE
+  $ io-ops bonuses-bv:redeem-bonuses
+
+OPTIONS
+  -c, --containerFolder=containerFolder  (required) container folder
```
[the remark this answers is not in the document]
Comment (pasqualedevita) [reply README.md]: renamed in day

File: README.md [hunk 8]
```diff
@@ -199,6 +200,25 @@ EXAMPLE
...
+  -k, --apiKey=apiKey                    (required) api key
```
[the remark this answers is not in the document]
Comment (pasqualedevita) [reply README.md]: fixed description

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 5 above)
Comment (pasqualedevita) [reply src/commands/bonuses-bv/redeem-bonuses.ts]: setted to 10s

File: README.md [hunk 9]
```diff
@@ -199,6 +200,25 @@ EXAMPLE
...
+  -o, --tmpDir=tmpDir                    (required) [default: tmp] temp directory
```
[the remark this answers is not in the document]
Comment (pasqualedevita) [reply README.md]: removed required attribute when it's setted a default value.

[the remark this answers is not in the document]
Comment (pasqualedevita) [reply README.md]: We can see the file without go on Azure and search it, but it is not a big benefit.
In the new version, the script reprocess the request on the fly without save it in temp directory.
Thank you :)

File: src/utils/azure.ts (hunk 6 above)
Comment (pasqualedevita) [reply src/utils/azure.ts]: removed

**Comment by pasqualedevita** [comment]
added flags fromTime and toTime to filter redeemed requests

File: README.md [hunk 10]
```diff
@@ -199,6 +200,28 @@ EXAMPLE
...
+
+```
+USAGE
+  $ io-ops bonuses-bv:redeem-bonuses
+
+OPTIONS
+  --apiKey=apiKey      (required) bonus vacanze api key
+  --apiUrl=apiUrl      [default: https://api.io.italia.it/api/bonus-vacanze/v1/redeemed] bonus vacanze api url
+
+  --day=day            (required) filter reddem request from specified day (utc timezome, required format yyyy-MM-dd, ie
```
Comment (Undermaken) [inline README.md:215]:
```suggestion
  --day=day            (required) filter redeemed request from specified day (utc timezone, required format yyyy-MM-dd, ie
```

File: src/commands/bonuses-bv/redeem-bonuses.ts [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+
+export default class RedeemBonuses extends Command {
+  public static description = "Reprocess redeemed bonuses in the specified day";
+
+  public static args = [];
+
+  public static flags = {
+    day: flags.string({
+      description:
+        "filter reddem request from specified day (utc timezome, required format yyyy-MM-dd, ie 2020-05-25)",
```
Comment (Undermaken) [inline src/commands/bonuses-bv/redeem-bonuses.ts:21]:
```suggestion
        "filter redeemed request from specified day (utc timezone, required format yyyy-MM-dd, ie 2020-05-25)",
```

File: src/commands/bonuses-bv/redeem-bonuses.ts [hunk 12]
```diff
@@ -0,0 +1,202 @@
...
+      description: "bonus vacanze api key",
+      required: true
+    })
+  };
+
+  public async run(): Promise<void> {
+    const { flags: parsedFlags } = this.parse(RedeemBonuses);
+
+    const day = DateTime.fromFormat(
+      parsedFlags.day + " " + "utc",
```
Comment (Undermaken) [inline src/commands/bonuses-bv/redeem-bonuses.ts:46]:
```suggestion
      parsedFlags.day + " utc",
```

File: src/commands/bonuses-bv/redeem-bonuses.ts [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+    );
+    if (!day.isValid) {
+      this.error("day is not valid");
+      return;
+    }
+
+    let fromTime = undefined;
+    if (parsedFlags.fromTime !== undefined) {
+      fromTime = DateTime.fromFormat(
+        parsedFlags.day + " " + parsedFlags.fromTime + " " + "utc",
```
Comment (Undermaken) [inline src/commands/bonuses-bv/redeem-bonuses.ts:57]:
```suggestion
        `${parsedFlags.day} ${parsedFlags.fromTime} utc`,
```
[it's the same](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) but it's more readable

**Review by Undermaken — CHANGES_REQUESTED** [review]

File: README.md (hunk 10 above)
Comment (pasqualedevita) [reply README.md:215]: fixed

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 12 above)
Comment (pasqualedevita) [reply src/commands/bonuses-bv/redeem-bonuses.ts]: fixed

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 13 above)
Comment (pasqualedevita) [reply src/commands/bonuses-bv/redeem-bonuses.ts]: fixed, thank you!

File: src/commands/bonuses-bv/redeem-bonuses.ts (hunk 11 above)
Comment (pasqualedevita) [reply src/commands/bonuses-bv/redeem-bonuses.ts]: fixed

**Comment by pasqualedevita** [comment]
Reprocess redeemed bonuses finished. If it's ok we can merge into master
